crates/red_knot_python_semantic/src/semantic_index/use_def.rs

```diff
@@ -282,48 +282,61 @@ impl<'db> UseDefMapBuilder<'db> {
         // It's possible we may end up with some old entries in `all_definitions` that nobody is
         // pointing to, but that's OK.
 
-        for (symbol_id, to_merge) in snapshot.definitions_by_symbol.iter_enumerated() {
-            let current = &mut self.definitions_by_symbol[symbol_id];
+        // We never remove symbols from `definitions_by_symbol` (it's an IndexVec, and the symbol
+        // IDs must line up), so the current number of known symbols must always be equal to or
+        // greater than the number of known symbols in a previously-taken snapshot.
+        debug_assert!(self.definitions_by_symbol.len() >= snapshot.definitions_by_symbol.len());
+
+        for (symbol_id, current) in self.definitions_by_symbol.iter_mut_enumerated() {
+            let Some(to_merge) = snapshot.definitions_by_symbol.get(symbol_id) else {
+                // Symbol not present in snapshot, so it's unbound from that path.
+                current.may_be_unbound = true;
+                continue;
+            };
 
             // If the symbol can be unbound in either predecessor, it can be unbound post-merge.
             current.may_be_unbound |= to_merge.may_be_unbound;
 
             // Merge the definition ranges.
-            if current.definitions_range == to_merge.definitions_range {
-                // Ranges already identical, nothing to do!
-            } else if current.definitions_range.end == to_merge.definitions_range.start {
-                // Ranges are adjacent (`current` first), just merge them into one range.
-                current.definitions_range =
-                    (current.definitions_range.start)..(to_merge.definitions_range.end);
-            } else if current.definitions_range.start == to_merge.definitions_range.end {
-                // Ranges are adjacent (`to_merge` first), just merge them into one range.
+            let cur = &mut current.definitions_range;
```

Comment:
Nit: (and `snapshot`?)
```suggestion
            let current = &mut current.definitions_range;
```

Reply:
I don't like shadowing, I find it confusing :/ I'll try to find a solution that doesn't involve either shadowing or abbreviations :)

Reply:
Ok, I got over it and just shadowed. I think it's not too bad: we shift gears to merging the definition ranges, and from that point on `current` and `snapshot` refer to just the ranges.

Reply:
I disliked shadowing in JS but it's more common in the Rust world and I kind of got used to it.